sql: remove Impure flag

RaduBerinde opened the pull request:
This set of commits changes all uses of Impure to use volatilities instead, and removes the flag altogether.

#### sql: use volatility when sanitizing expressions

As part of work toward removing the coarse (and unreliable) `Impure` flag, this
change modifies the expression sanitizers to take a `maxVolatility` argument
instead of an `allowImpure` boolean. Currently we only support either Stable or
Volatile as max-volatility. There are some TODO cases where we may want to
support Immutable (but note that we would have to check operators and casts as
well).

Release note: None

#### sql: use volatility in normalization / IsConst

Release note: None

#### sqlsmith: use volatility in sqlsmith check

Update the impure check in sqlsmith to check the Volatility. We want only
immutable functions so the results of any given query are consistent across
servers.

Release note: None

#### sql: remove Impure flag

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50678)
<!-- Reviewable:end -->

jordanlewis reviewed (approved):
> Reviewed 14 of 14 files at r1, 3 of 3 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50678)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @pbardea, @RaduBerinde, and @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/computed, line 295 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50678#-MAizD8e4OiMtoKVbeCD:-MAizD8e4OiMtoKVbeCE:bvkng86) ([raw file](https://github.com/cockroachdb/cockroach/blob/1caa95eb1644bad9c07a0004c33a84fbc67cf94c/pkg/sql/logictest/testdata/logic_test/computed#L295)):*
> ```
> statement ok
> CREATE TABLE y_stable (
>   a STRING AS (now()::STRING) STORED
> ```

Is this a desirable behavior change? What does it mean to have now() in a computed column - when is it evaluated?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50678)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @pbardea, and @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/computed, line 295 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50678#-MAizD8e4OiMtoKVbeCD:-MAj-cOT50NHMJjxGJ3L:bxpd89q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1caa95eb1644bad9c07a0004c33a84fbc67cf94c/pkg/sql/logictest/testdata/logic_test/computed#L295)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Is this a desirable behavior change? What does it mean to have now() in a computed column - when is it evaluated?
</blockquote></details>

Good question. I figured it's calculated whenever we insert, and whenever we update values it depends on (but there's probably some problems with that).

I am open to restricting to immutable (though we will need to check casts and operators too, I would leave that for a future change). I was afraid that stuff will stop working (we have a bunch of stable things that weren't impure).


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
There is also some fallout for partitioning I don't know what to do about. Using constant dates and timestamp strings involves non-immutable casts (they can depend on the current timezone), and we error out with "partition values must be constant". I will look into perhaps getting more granular with these casts and identify a subset that does not actually depend on context.

RaduBerinde committed:
sqlsmith: use volatility in sqlsmith check

Update the impure check in sqlsmith to check the Volatility. We want only
immutable functions so the results of any given query are consistent across
servers.

Release note: None

RaduBerinde committed:
sql: use volatility in normalization / IsConst

Release note: None

RaduBerinde committed:
sql: use volatility when sanitizing expressions

As part of work toward removing the coarse (and unreliable) `Impure` flag, this
change modifies the expression sanitizers to take a `maxVolatility` argument
instead of an `allowImpure` boolean. Note that currently we only check
functions, we will need to check operators and casts for the cases where we
require Immutable expressions (this is left as a TODO).

As part of this commit, we also correct the volatility of `concat`, `concat_ws`,
`convert_to`, `convert_from`. This is easy to verify, as the implementations
don't use the eval context. Note that the `concat` in Postgres can convert to
string any argument (which is the non-immutable part), but our function only
takes strings (which explains the difference).

Release note: None

RaduBerinde committed:
sql: remove Impure flag

Release note: None

RaduBerinde commented:
All right, I improved the way the constant casts work (which allows partitioning to work on timestamps and such) and I switched to requiring Immutable for computed columns.

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 10 of 14 files at r1, 4 of 9 files at r4, 4 of 4 files at r5, 1 of 1 files at r6, 8 of 8 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50678#-:-MB0z98nCz_TbYnVLy2X:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @pbardea, @RaduBerinde, and @rohany)

___
*[pkg/sql/sqlbase/table.go, line 66 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/50678#-MB0yj3Y2It8WafLflk5:-MB0yj3Y2It8WafLflk6:bw7etg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/126451bf658b63d76730bd63783a607ed7c62b62/pkg/sql/sqlbase/table.go#L66)):*
> ```Go
> 		flags |= tree.RejectVolatileFunctions
> 
> 	case tree.VolatilityVolatile:
> ```

[nit] I'm not sure it's necessary, but a small comment explaining why this is a no-op might prevent confusion or someone thinking this line should be removed.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
CC @rohany for the change in `TestBackupRestoreUserDefinedTypes` (backup_test.go)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50678#-:-MB1NiuY3CIWMWsY9Nd1:bexqg9m)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @mgartner, @pbardea, and @rohany)



<!-- Sent from Reviewable.io -->

rohany commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -1350,8 +1350,9 @@ INSERT INTO d.t2 VALUES (ARRAY['howdy']), (ARRAY['hi']);
 CREATE TABLE d.expr (
 	x d.greeting,
   y d.greeting DEFAULT 'hello',
-	z d.greeting[] AS (enum_range(x, 'hi')) STORED,
```
Why did this have to change? `enum_range` is a `VolatilityStable` function

RaduBerinde reviewed (commented):
Jordan argued above that we shouldn't allow stable functions (like now()) in computed columns.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50678#-:-MB1PqgBBuRjgnkSCCLy:bihgcyq)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @mgartner, @pbardea, and @rohany)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
